GNUmakefile

```diff
@@ -99,11 +99,14 @@ go-solidity-wrappers-ocr2vrf: pnpmdep abigen ## Recompiles solidity contracts an
 generate: abigen codecgen mockery ## Execute all go:generate commands.
 	go generate -x ./...
 
+# NOTE: When running the below command on OSX, you must have the GNU version of `find`.
+# Install this via `brew install findutils`, and follow the instructions about adding it
+# to your PATH.
 .PHONY: testscripts
 testscripts: chainlink ## Install and run testscript against testdata/scripts/* files.
 	go install github.com/rogpeppe/go-internal/cmd/testscript@latest
 	find testdata/scripts -type d | xargs -I % \
-	sh -c 'ls %/*.txtar > /dev/null 2>&1 || return 0 && PATH=$(CURDIR):$(PATH) testscript -e CL_DEV=true -e COMMIT_SHA=$(COMMIT_SHA) -e VERSION=$(VERSION) $(TS_FLAGS) %/*.txtar'
+	sh -c 'ls %/*.txtar > /dev/null 2>&1 || exit 0 && PATH=$(CURDIR):$(PATH) testscript -e CL_DEV=true -e COMMIT_SHA=$(COMMIT_SHA) -e VERSION=$(VERSION) $(TS_FLAGS) %/*.txtar'
```

Comment:
Hmm, I don't think anything after the `exit 0` will actually run, right?

Reply:
Yeah this is by design - we want to skip the repos that don't have any `*.txtar` files. Just trying to make things more portable here since what I wrote didn't work on OSX.

Reply:
That took be a minute to comprehend b/c of the inlining, WDYT of this?
Then we can drop the dependency on `find` too. 

FYI, I haven't tested this.
```sh
testscripts: chainlink
	go install github.com/rogpeppe/go-internal/cmd/testscript@latest
	for dir in testdata/scripts/*/; do \
		files=$$(ls $$dir*.txtar 2> /dev/null); \
		[ ! -z "$$files" ] && PATH=$(CURDIR):$(PATH) testscript -e CL_DEV=true -e COMMIT_SHA=$(COMMIT_SHA) -e VERSION=$(VERSION) $(TS_FLAGS) $$files; \
	done

```

Reply:
Formatted version @jmank88 @chainchad 
```sh
testscripts: chainlink
	go install github.com/rogpeppe/go-internal/cmd/testscript@latest
	for dir in testdata/scripts/*/; do \
		files=$$(ls $$dir*.txtar 2> /dev/null); \
		if [ ! -z "$$files" ]; then \
			PATH=$(CURDIR):$(PATH) testscript \
				-e CL_DEV=true \
				-e COMMIT_SHA=$(COMMIT_SHA) \
				-e VERSION=$(VERSION) \
				$(TS_FLAGS) \
				$$files; \
		fi; \
	done

```

Reply:
That's some intense bash 😅 We should test it but if it works, it's nice to avoid worrying about which flavor of `find` is on a system.

Might just do `if [ -n "$$files" ]; then` though.

Reply:
its pretty much the same amount of logic, i just split the args across new lines so its not a giant one liner + used an `if` rather than (ab)using short circuiting :laughing:

Reply:
Awesome! Bonus points for `&`/`wait` parallelism?

Reply:
Hmmm... I've been trying to adapt your solution @HenryNguyen5 but am struggling to get the glob to return all of the subdirectories recursively, which is what `find` was doing for us.

The command I have is:

```
	for dir in testdata/**/*/; do \
		files=$$(ls $$dir*.txtar 2> /dev/null); \
		if [ "$$files" ]; then \
			PATH=$(CURDIR):$(PATH) testscript \
			-e CL_DEV=true \
			-e COMMIT_SHA=$(COMMIT_SHA) \
			-e VERSION=$(VERSION) \
			$$files; \
		fi; \
	done
```

which works in my (zsh-)shell but not using `make` since `make` will use `sh` by default.

The problematic line is:

```
for dir in testdata/**/*/; do echo $dir; done
```

which returns all of the right directories in my shell, but not via make since sh doesn't support `**`.

```
sh-3.2$ for dir in testdata/**/*/; do echo $dir; done
testdata/scripts/node/
```

(Changing the make shell to bash results in the same problem -- 
`bash-3.2$ for dir in testdata/**/*/; do echo $dir; done`

Stackoverflow tells me that recursive globs are supported in version of bash > 4, so maybe we could just print out an error if the bash version used is too old (3.2, which comes default on OSX is from 2007!)?

Reply:
This version uses find, but avoids the GNU extension we had earlier, hopefully we get the best of both worlds now. LMK if this works.

```
testscripts: chainlink
	go install github.com/rogpeppe/go-internal/cmd/testscript@latest
	
	# For every file path that is separated by a null character
	while IFS= read -r -d '' file; do \
		PATH=$(CURDIR):$(PATH) testscript \
			-e CL_DEV=true \
			-e COMMIT_SHA=$(COMMIT_SHA) \
			-e VERSION=$(VERSION) \
			$(TS_FLAGS) \
			"$$file"; \
	# Read file paths from the output of the find command
	done < <(find testdata/scripts -type f -name "*.txtar" -print0)
```

Reply:
:mage_man:

Reply:
Wait is this calling `testscript` on each individual `.txtar` file? We only _need_ to call it on each directory (that has any `.txtar`), though I'm not sure if it matters :thinking:

Reply:
Ah i see, this script has become quite a bit just to remove the system dep heh.

Reply:
I created a go version, you can run it with `-dryrun` to see what it'd execute.
```go
import (
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func main() {
	dryrun := flag.Bool("dryrun", false, "Enable dry run mode to print commands without executing them")
	flag.Parse()

	root := "testdata/scripts"
	processedDirs := make(map[string]bool)

	filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !info.IsDir() && strings.HasSuffix(info.Name(), ".txtar") {
			dir := filepath.Dir(path)

			// Skip processing if the directory has already been processed
			if processedDirs[dir] {
				return nil
			}

			// Mark the directory as processed
			processedDirs[dir] = true

			cmd := exec.Command("testscript",
				"-e", "CL_DEV=true",
				"-e", fmt.Sprintf("COMMIT_SHA=%s", os.Getenv("COMMIT_SHA")),
				"-e", fmt.Sprintf("VERSION=%s", os.Getenv("VERSION")),
				"-e", fmt.Sprintf("TS_FLAGS=%s", os.Getenv("TS_FLAGS")),
				dir,
			)

			cmd.Env = append(os.Environ(), fmt.Sprintf("PATH=%s:%s", os.Getenv("CURDIR"), os.Getenv("PATH")))

			if *dryrun {
				fmt.Printf("DRY RUN: testscript command: %v\n", cmd.Args)
			} else {
				cmd.Stdout = os.Stdout
				cmd.Stderr = os.Stderr

				if err := cmd.Run(); err != nil {
					fmt.Printf("Error running testscript for %s: %v\n", dir, err)
				}
			}
		}
		return nil
	})
}
```

Reply:
It might be simpler to submit a patch upstream to add a recursive mode/filter to testscript itself :laughing:

Reply:
Thanks!

@HenryNguyen5 I liked your idea of just using go here, so I created a command to list all .txtar files under testdata/scripts largely based off what you wrote above.

Reply:
lol but we already have a go version internally. This was meant to replicate that behavior for the command line.
We should at least share the code between the two, but I'm worried this is becoming a mess.

Reply:
@jmank88 Ah I see now... Could we not just run that test in isolation in `make testscripts`? eg. `go test main_test.go`.?

Reply:
The point is to be able to invoke the command line `testscript` program. This is fundamentally different than running the internal version. It allows us to use the update (-u) flag, debug more easily, etc. Later we can leverage it to run against any installed binary, instead of the current working state. I've been trying to rig up a docker variant too: basically `alias chailnik="docker run smartcontract/chainlink"` should pass all the same tests.

Reply:
Not sure if a good idea, but we could just have a go script that replaces the `find` functionality 😊